### PR TITLE
Cloudflare createdomain

### PR DESCRIPTION
**Pull request by tiernano**
fixing issue #112

### CODE REVIEW

File: normalize/validate.go [hunk 1]
```diff
@@ -359,4 +359,4 @@ func applyRecordTransforms(domain *models.DomainConfig) error {
 		}
 	}
 	return nil
-}
+}
```
Comment (captncraig) [inline normalize/validate.go]: Looks like you are on windows? Line endings are a pain in the ass. I always have to set my editors and git config up to not touch line endings at all for my go projects.

File: providers/diff/diff.go [hunk 2]
```diff
@@ -175,4 +175,4 @@ func sortedKeys(m map[string]*models.RecordConfig) []string {
 	}
 	sort.Strings(s)
 	return s
-}
+}
```
Comment (captncraig) [inline providers/diff/diff.go]: a few different line ending issues. Gofmt will not pass this change.

File: providers/cloudflare/cloudflareProvider.go [hunk 3]
```diff
@@ -103,6 +103,9 @@ func (c *CloudflareApi) GetDomainCorrections(dc *models.DomainConfig) ([]*models
 	corrections := []*models.Correction{}
 
 	for _, d := range del {
+		if dc.KeepUnknown {
```
Comment (captncraig) [inline providers/cloudflare/cloudflareProvider.go]: This part is not needed anymore since #111

File: providers/cloudflare/cloudflareProvider.go [hunk 4]
```diff
@@ -300,3 +303,24 @@ func getProxyMetadata(r *models.RecordConfig) map[string]string {
 		"proxy": fmt.Sprint(proxied),
 	}
 }
+
+func (c *CloudflareApi) EnsureDomainExists(domain string) error {
+	err := c.fetchDomainList()
+
```
Comment (captncraig) [inline providers/cloudflare/cloudflareProvider.go]: A bit of unnecessary white space here. Let's condense the empty lines a bit.

File: providers/cloudflare/cloudflareProvider.go [hunk 5]
```diff
@@ -300,3 +303,24 @@ func getProxyMetadata(r *models.RecordConfig) map[string]string {
...
+		return err
+	}
+	if _, ok := c.domainIndex[domain]; ok {
+		return nil
+	}
+	fmt.Printf("Adding zone for %s to Cloudflare account\n", domain)
+
+	id, err = c.createZone(domain)
+
+	fmt.Printf("ID: %s", id)
```
Comment (captncraig) [inline providers/cloudflare/cloudflareProvider.go]: Don't need debug in final version.

File: providers/cloudflare/rest.go [hunk 6]
```diff
@@ -88,6 +88,32 @@ func (c *CloudflareApi) deleteRec(rec *cfRecord, domainID string) *models.Correc
 	}
 }
 
+func (c *CloudflareApi) createZone(domainName string) (string, error) {
+	type createZone struct {
+		Name string `json:"name"`
+	}
+	endpoint := zonesURL
```
Comment (captncraig) [inline providers/cloudflare/rest.go]: probably don't need to redeclare endpoint as a variable. Just use zonesURL the one place you use endpoint later.

Comment (captncraig) [inline providers/cloudflare/rest.go]: also condense whitespace here.

**Comment by tiernano** [comment]
Cool. Will make those fixes. Glad I Dident screw up too badly! ;-)

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by tiernano** [comment]
all requested fixes should have been made now... Any problems, shout.

File: main.go [hunk 7]
```diff
@@ -149,6 +149,8 @@ func main() {
 					if err != nil {
 						fmt.Printf("Error creating domain: %s\n", err)
 					}
+				} else {
```
Comment (captncraig) [inline main.go]: Remove this debug code now.

File: providers/cloudflare/cloudflareProvider.go [hunk 8]
```diff
@@ -227,6 +227,11 @@ func newCloudflare(m map[string]string, metadata json.RawMessage) (providers.DNS
 		return nil, fmt.Errorf("Cloudflare apikey and apiuser must be provided.")
 	}
 
+	err := c.fetchDomainList()
```
Comment (captncraig) [inline providers/cloudflare/cloudflareProvider.go]: Failing to compile. Should be
```
err := api.fetchDomainList()
	if err != nil {
		return nil, err
	}
```

File: providers/cloudflare/cloudflareProvider.go [hunk 9]
```diff
@@ -300,3 +305,17 @@ func getProxyMetadata(r *models.RecordConfig) map[string]string {
 		"proxy": fmt.Sprint(proxied),
 	}
 }
+
+func (c *CloudflareApi) EnsureDomainExists(domain string) error {
+	err := c.fetchDomainList()
```
Comment (captncraig) [inline providers/cloudflare/cloudflareProvider.go]: If we fetch the domain list when we create the provider, lets just store it and not need to fetch it all over the place again.

**Comment by tiernano** [comment]
making changes and will check-in in a few min... thanks!

**Comment by tiernano** [comment]
all requested changes have now been made.

**Comment by captncraig** [comment]
Thank you!

**Pull request merged**